Light History Store for full nodes

viquezclaudio opened the pull request:
Use the light history store for full node clients
Necessary changes to use the history interface
Several changes to the usage of the history store


...
#### This fixes #1704.

## Pull request checklist

- [ ] All tests pass. The project builds and runs.
- [ ] I have resolved any merge conflicts.
- [ ] I have resolved all `clippy` and `rustfmt` warnings.

viquezclaudio committed:
Light History Store for full nodes

Use the light history store for full node clients
Necessary changes to use the history interface
Several changes to usage of the history store

sisou commented:
Does this fix #1704?

viquezclaudio commented:
Correct, this is the proper fix for 1704.
However this still needs more testing.

sisou commented:
:+1: I updated the PR description with that issue number.

paberr commented on blockchain/src/blockchain/history_sync.rs:
```diff
@@ -96,24 +97,41 @@ impl Blockchain {
+            } else {
+                let new_txns = vec![txn.clone()];
+                txns_per_block.insert(bn, new_txns);
+            }
```
```suggestion
            txns_per_block.entry(bn).or_default().push(txn.clone());
```

paberr commented on blockchain/src/blockchain/history_sync.rs:
```diff
@@ -96,24 +97,41 @@ impl Blockchain {
+        }
+
+        let mut root = Blake2bHash::default();
+        for bn in txns_per_block.keys().sorted() {
+            let hist_txs = txns_per_block.get(bn).unwrap();
```
Using a `BTreeMap`, the keys would already be sorted.
Then, this could be simplified to `for (bn, hist_txs) in txns_per_block`.
Not sure whether it's worth it though.

Also, there's an inconsistency in naming: `txs` vs `txns`

viquezclaudio replied:
Yes, I considered using a BTreeMap, but I wasn't sure if it was worth to add the dependency and use it vs just sorting the keys

hrxi replied:
> if it was worth to add the dependency

It's in the standard library, so would be no additional dependency: [`std::collections::BTreeMap`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html).

paberr commented on blockchain/src/history/history_store.rs:
```diff
@@ -648,13 +655,38 @@ impl HistoryInterface for HistoryStore {
+
+        for hist_tx in ext_hash_vec {
+            // If the transaction is inside the validity window, return true.
+            if hist_tx.block_number >= validity_window_start {
```
Currently, we only check whether the transaction occurred *after* the validity window started.
However, it could potentially still fall outside the window if it occurred after it finished.
The second boundary check seems missing.

viquezclaudio replied:
Here I'm not doing anything different than what we were doing before, i.e: I just moved this code: https://github.com/nimiq/core-rs-albatross/blob/albatross/blockchain/src/blockchain/wrappers.rs#L196
to the history store file.

Do you suggest that we used to have a bug? 
What do you mean by "if it occurred after it finished." ?

nibhar replied:
I don't know what that constant is, but that would be the check. This does not make any assumptions on where the chain is so the validity window that starts at `validity_window_start` could be a long time ago. Thus there are transactions after this window which would be valid.

i.e.
```suggestion
            if hist_tx.block_number >= validity_window_start && hist_tx.block_number <= validity_window_start + VALIDITY_WINDOW {
```

Depending on how it was used this would potentially not be buggy, but checking both boundaries is certainly more sound.

paberr commented on blockchain/src/history/mmr_store.rs:
```diff
@@ -190,6 +191,14 @@ impl<'a, 'env> LightMMRStore<'a, 'env> {
     ) -> Self {
         let size = get_size(hist_tree_table, tx, block_number);
+
+        // If size is 0 we need to copy the entries from the previous block number to the new one
+        if size == 0 {
+            init_mmr_from_block_number(hist_tree_table, tx, block_number);
+        }
```
I don't understand what this is about

viquezclaudio replied:
For the light history store, what we are using are "peaks-only MMRs" 
However, this means that if we have to revert blocks, we don't have enough information to just remove txns and go back to a previous state, as we do with the regular MMR.

So what we are doing is to store "peaks only mmrs" per block, i.e. essentially we have "snapshots" of how the peaks only MMR looked at each block. 

However, since history nodes construct a big MMR with all the transactions of the given epoch, and this is the root that we need to verify against, then it means the starting point for block N is the peaks only mmr of block N-1, which is what this line is doing:
when we start pushing transactions for a block that we didn't have before, we first need to start with the peaks only mmr of the previous block and continue from there.

Let me know if this clarifies just question

paberr commented on blockchain/src/history/mmr_store.rs:
```diff
@@ -232,6 +241,72 @@ pub fn remove_block_from_store(
+        };
+
+        // Deconstruct the key into a block number and a node index.
+        let (new_block_number, _) = key_to_index(next_key).unwrap();
```
`key_to_index` returns an `epoch_number` instead of a `block_number` (according to the variable names in the function; so this needs updating)

viquezclaudio replied:
I'll update the comment, because we are using the same functions that the regular history store uses (which constructs keys based on {epoch_number, index} , but for the peaks only mmr, we construct keys based on {block_number, index}

paberr commented on blockchain/src/history/mmr_store.rs:
```diff
@@ -232,6 +241,72 @@ pub fn remove_block_from_store(
 
+/// Copies the mmr entries of the previous block number
+/// This is used to initialize the tree for the given block number, using the previous tree as a baseline
+pub fn init_mmr_from_block_number(
```
What is the advantage of copying the contents over accessing the previous ones when needed?
Less complexity when accessing?

viquezclaudio replied:
See my comment above.

viquezclaudio committed:
Use the database for the validity store

Mantain the txns in the validity store in database tables.

viquezclaudio committed:
Use BTreeMap when pushing txns during validity sync

viquezclaudio committed:
Fix history store range function

The history store range function now uses a database transaction
Added unittest for total_len_at_epoch in the light history store

viquezclaudio committed:
Test transaction_in_validity_window

Add an unittest to exercise the transaction_in_validity_window functionality in the light history store
Fixed typo in range inclusion when pruning the validity store

nibhar commented on blockchain/src/blockchain/blockchain.rs:
```diff
@@ -69,13 +70,17 @@ pub struct BlockchainConfig {
     pub max_epochs_stored: u32,
+    /// The history store that is used by the full blockchain.
+    /// If this is set to true, the light history store is used.
+    pub light_history_store: bool,
```
Is this not already somewhat encoded in `keep_history` flag? Is there a usecase for `keep_history & light_history_store`? Similarly is there a usecase for `!use_history & !light_history_store`?

viquezclaudio replied:
AFAIK, `keep_history` is used to determine if we want to prune the history store,  keeping only up to "max_epochs_stored" epochs, whereas "light_history_store" specifies the type of History Store that is being used, and these two are independent.
i.e.: you could potentially have a node using a light story store where keep_history is true  (not sure if there is a  use case for this)
Also, with this implementation, it is possible to have full nodes with the regular history store and full nodes with the light history store.

However, if we decide that:

- Full nodes only use the light story store and they only keep up to max_epochs_stored 
 
Then this can indeed be simplified, as the different combinations mentioned above would no longer be supported

nibhar replied:
Fair enough. More flexibility is good, if it does not lead to configurations which are bogus. Thanks for explaining.

nibhar commented on blockchain/src/blockchain/history_sync.rs:
```diff
@@ -96,24 +96,39 @@ impl Blockchain {
-        // Store the new historic transactions into the History tree.
-        if let Some((root, _)) = this.history_store.add_to_history(&mut txn, epoch, history) {
-            txn.commit();
+        //let mut txns_per_block: HashMap<u32, Vec<HistoricTransaction>> = HashMap::new();
```
Leftover?

nibhar commented on blockchain/src/blockchain/history_sync.rs:
```diff
@@ -96,24 +96,39 @@ impl Blockchain {
         this: RwLockUpgradableReadGuard<Blockchain>,
-        epoch: u32,
         history: &[HistoricTransaction],
     ) -> Option<Blake2bHash> {
```
Looks to me as if the changes made this no longer fail for any reason, but panic instead.
```suggestion
    ) -> Blake2bHash {
```

nibhar commented on blockchain/src/history/history_store.rs:
```diff
@@ -1156,15 +1188,100 @@ impl HistoryInterface for HistoryStore {
-        todo!()
+        match block {
+            nimiq_block::Block::Macro(macro_block) => {
+                // Store the transactions and the inherents into the History tree.
```
The transactions Vec of the following call is empty. Is this a misleading comment or are transactions missing here?

viquezclaudio replied:
I will update the comment, we are only storing inherent for this particular case

nibhar commented on blockchain/src/history/light_history_store.rs:
```diff
@@ -85,25 +93,44 @@ impl HistoryInterface for LightHistoryStore {
+        self.validity_store
+            .delete_block_transactions(txn, block.block_number());
+
+        // TODO: We dont keep track of the size of the txns that we removed, is this necessary?
```
TODO

`Some(0)` also seems to be incorrect, so I would suggest to either change the return type, or to actually return the correct return value.

viquezclaudio replied:
The problem is that this comes from the regular history store, where we have actual transactions and the pruning process is different.
For the light history store, internally, we just keep peaks only MMRs so we don't have the information to calculate this, so not sure what we want to do here

viquezclaudio replied:
Created #2593 to track this issue

nibhar commented on blockchain/src/history/light_history_store.rs:
```diff
@@ -147,28 +203,50 @@ impl HistoryInterface for LightHistoryStore {
-        tree.len()
+        let (first_bn, last_bn) = self.history_store_range(Some(txn));
+
+        let length = if Policy::epoch_at(last_bn) == epoch_number {
+            // Get peaks mmr of the latest block stored
+            let tree = PeaksMerkleMountainRange::new(LightMMRStore::with_read_transaction(
+                &self.hist_tree_table,
+                txn,
+                last_bn,
+            ));
+
+            // Get the Merkle tree length
+            tree.num_leaves()
+        } else {
+            if first_bn < (Policy::first_block_of(epoch_number + 1).unwrap() - 1) {
+                // Get peaks mmr of the latest block stored
+                let tree = PeaksMerkleMountainRange::new(LightMMRStore::with_read_transaction(
+                    &self.hist_tree_table,
+                    txn,
+                    Policy::first_block_of(epoch_number + 1).unwrap() - 1,
+                ));
+
+                // Get the Merkle tree length
+                tree.num_leaves()
+            } else {
+                log::debug!(" validity out of bounds!");
+                0
+            }
+        };
```
The comments here are not helpful to me, but I would have appreciated some helpful comments.

I don't know how much it matters and if I understand correctly but when epochs can be incomplete in the `hist_tree_table` then neither of these return the actual size of the epoch, but only the known size.

viquezclaudio replied:
I will update the comments

nibhar commented on blockchain/src/history/light_history_store.rs:
```diff
@@ -147,28 +203,50 @@ impl HistoryInterface for LightHistoryStore {
+            }
+        };
+
+        length
```
Could be returned directly from the conditionals. Conditional can be collapsed.

nibhar commented on blockchain/src/history/light_history_store.rs:
```diff
@@ -177,6 +255,13 @@ impl HistoryInterface for LightHistoryStore {
+        self.validity_store.update_validity_store(txn, block_number);
+
         // Return the history root.
         Some((root, 0))
```
That `0` also seems to be incorrect. Either change the return type, or return the actual correct value.

viquezclaudio replied:
Same problem as before, we dont keep track of the sizes in the light history store

nibhar commented on blockchain/src/history/mmr_store.rs:
```diff
@@ -232,6 +262,75 @@ pub fn remove_block_from_store(
 
+/// Copies the mmr entries of the previous block number
+/// This is used to initialize the tree for the given block number, using the previous tree as a baseline
+pub fn init_mmr_from_block_number(
```
Not entirely sure why the MMR of a single block preceding the current one is the way to go when the MMR actually spans an entire epoch. A comment explaining this would be nice as it certainly threw me of and I cannot see mention as to why in this PR.

viquezclaudio replied:
I will update the comment

nibhar commented on consensus/src/sync/light/validity_window.rs:
```diff
@@ -51,17 +51,71 @@ impl<TNetwork: Network> LightMacroSync<TNetwork> {
+            let mut chunk_index = 0;
+            let mut root_hash = expected_root.clone();
+            let mut last_chunk_items: Option<usize> = None;
+            let mut verifier_block_number = verifier_block_number;
+            let mut election_in_window = election_in_window;
```
These could just be `mut` in the arguments of the function already.

nibhar commented on consensus/src/sync/light/validity_window.rs:
```diff
@@ -51,17 +51,71 @@ impl<TNetwork: Network> LightMacroSync<TNetwork> {
+
+                    let (first_bn, last_bn) = blockchain_wr.history_store.history_store_range(None);
+
+                    // This means we already know the first epoch so we need to request the missing items from the current epoch
```
What does the first epoch have to do with that? I assume you mean the epoch in which the start of the window is. Reading this and trying to understand it I do not necessarily know yet there is at most 2 relevant epochs.

viquezclaudio replied:
Yes, the comment is misleading. I will update it

nibhar commented on consensus/src/sync/light/validity_window.rs:
```diff
@@ -51,17 +51,71 @@ impl<TNetwork: Network> LightMacroSync<TNetwork> {
+                        root_hash = blockchain_wr.macro_head().header.history_root.clone();
+                        verifier_block_number = blockchain_wr.macro_head().block_number();
+                        election_in_window = false;
+                    } else if last_bn >= validity_window_start {
```
```suggestion
                    } else if last_bn >= validity_window_start {
                    // The window is not empty but there are missing items in the epoch the window starts in.
```
Or similar.

nibhar commented on network-interface/src/peer_info.rs:
```diff
@@ -66,9 +66,7 @@ impl Services {
-            NodeType::Full => {
-                Services::ACCOUNTS_PROOF | Services::FULL_BLOCKS | Services::ACCOUNTS_CHUNKS
-            }
+            NodeType::Full => Services::ACCOUNTS_PROOF | Services::FULL_BLOCKS,
```
Interesting. So accounts chunks can now only be received by History nodes?

viquezclaudio replied:
Accounts chunks can only be provided by history nodes.
They can no longer be provided by full nodes. 

This struct represents what each node type can provide

viquezclaudio replied:
As per our discussion, I created a new issue for this #2592

nibhar reviewed (commented):
Very high level I am not a big fan of `bn` as `block_number`. I would much prefer to not abbreviate that to stay consistent throughout the codebase.

Generally looks mostly good to me. Left a couple of comments which I think are worth addressing. I personally would also appreciate some more documentation as to why things are done the way they are done, but that is just me.

A couple of clippy warnings need fixing too.

viquezclaudio committed:
Applied PR comments

nibhar reviewed (approved):
Good to go from my end even though I still don't like the `bn` abbreviations. I guess @paberr should take another look.
